Implemented Basic EKS Integration

ferruzzi opened the pull request:
Added hooks and operators for initial EKS support 

Includes Hooks and operators for the following EKS endpoints, along with tests, and documentation:
 - Create Cluster
 - Create Nodegroup
 - Delete Cluster
 - Delete Nodegroup
 -  Run Pod
 - ~Describe Cluster~
 - ~Describe Nodegroup~
 - ~List Cluster~
 - ~List Nodegroup~

 
 Increases the required version for the `moto` package from 2.0.8 to 2.0.10 in order to ingest the [new EKS support](https://github.com/spulec/moto/commit/61e2204941f3a7c6de831b71ea91c7eccc2ae645).

closes: #8544


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ferruzzi committed:
Implemented Basic EKS Integration

ashb commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -347,6 +347,7 @@ def __init__(
         client_type: Optional[str] = None,
         resource_type: Optional[str] = None,
         config: Optional[Config] = None,
+        **kwargs,
```
If we don't _do_ anything with these kwargs why do we need to accept them here?

ferruzzi replied:
That was being used to swallow the unused args when the Operators were creating the hook in their constructors.  I will remove it when I correct those.

ashb commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,346 @@
+        :param resourcesVpcConfig: The VPC configuration used by the cluster control plane.
+        :type resourcesVpcConfig: Dict
+
+        :return:  A JSON serialized string of the API call results.
```
Why return it as a string, not the Python object/dict?

ashb replied:
And follow-up question: why does `list_clusters` not follow this pattern? (it returns a dict)

ferruzzi replied:
Would you believe I had a Really Good Reason for it, but I can't remember why now?  It isn't how the other hooks are handled so I must have had a reason at the time, but I can't think of why.  I'll switch these to all return Dicts like the rest of the AWS/boto hooks.

ashb replied:
🤣

ashb commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,346 @@
+        kwargs["client_type"] = self.client_type
+        super().__init__(*args, **kwargs)
+
+    def create_cluster(self, name: str, roleArn: str, resourcesVpcConfig: Dict, **kwargs) -> str:
```
Interesting conundrum here. The Python naming style says these args should be `role_arn` and `resources_vpc_config` -- but _maybe_ it makes sense to mirror the Boto/AWS API names here.

What do other think, and what do we do in other AWS hooks?

ferruzzi replied:
My thinking, for what it is worth, was that the Hooks should mirror the API as closely as possible.  That would improve the learning curve for those coming to Airflow with some boto experience, and also make it easier for those new to the boto API and looking at their docs to see what is going on since the names match up.  

Existing AWS hooks are a mixed bag, but do lean towards snake_case.

mik-laj replied:
I think it is worth following the conventions used by boto to avoid any inaccuracies or mistake.

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,646 @@
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.eks_hook = EKSHook(**kwargs)
```
Plese don't create hooks in Operator constructors -- we generally try to avoid it as it _can_ end up creating network requests during DAG parse time which we want to avoid.

ferruzzi replied:
AHA.  Is that what those error messages were getting at in the CI tests that are failing?

ashb reviewed (commented):
Just a first pass -- haven't reviewed in depth at all yet.

ashb commented on tests/providers/amazon/aws/hooks/test_eks.py:
```diff
@@ -0,0 +1,872 @@
+    return _execute
+
+
+@unittest.skipIf(mock_eks is None, reason=PACKAGE_NOT_PRESENT_MSG)
```
Since this is not a unittest subclass
```suggestion
@pytest.mark.skipif(mock_eks is None, reason=PACKAGE_NOT_PRESENT_MSG)
```

ferruzzi replied:
Thanks, I'll incorporate that in my coming revision.

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,346 @@
+DEFAULT_PAGINATION_TOKEN = ''
+
+
+class EKSHook(AwsBaseHook):
```
EKSHook or EksHook? Which one do you prefer? Which is recommended by PEP? I asking, because it don't match to base class style.

ferruzzi replied:
Renamed to EksHook, you are right.  I like EKSHook better because it's an abbreviation but I'll follow conventions.

ferruzzi replied:
Fixed in 616d249

ashb replied:
Given it's an acronym keeping it as `EKSHook` is allowed, and just because we made the wrong decision in the past with AwsBaseHook doesn't mean we _have to_ stick by it here :)

mik-laj replied:
> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

https://www.python.org/dev/peps/pep-0008/

According to PEP-8, we should use `EKSHook`.

ferruzzi replied:
ACK.  Will revert this back to EKSHook and have more spine in the future :P

ferruzzi replied:
Reverted in https://github.com/apache/airflow/commit/667401ff7d1037edf52d17e9e84695edaee52cb5

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+
+    :param max_results: The maximum number of results to return.
+    :type max_results: int
+    :param next_token: The nextToken value returned from a previous paginated execution.
```
How is the user to pass this parameter?  Airflow does not support dynamic workflows, but the page count is dynamic.

ferruzzi replied:
I'm not sure I follow the question.   These are passed in from the DAG like any other parameter.

Many of the EKS APIs have a max result value of 100 entries so there should be a way to get the next paginated set.  They could get the nextToken through a prior task's XCOM or whatever, but I felt it should be available.

mik-laj replied:
When defining a DAG, the user should define all tasks. Here, however, there is a case where the number of tasks is not known until the first page is downloaded, and therefore cannot be defined in Airflow.

In the case of Google, we've always fetched all items from all pages. This way, the user could access all elements and did not have to define dynamic DAGs.

ferruzzi replied:
I see.  I implemented the hooks to follow the API as closely as possible.   To make sure I am on the same page before I change anything, you think it would be better for the hook to just return everything and not allow pagination at all?

ashb replied:
Generally the hook should handle the pagination/return a paginator (see the Athena and Glue hooks) and then the operator just iterates over it.

ferruzzi replied:
I think I understand what you are saying, thanks.  A comment below suggests that the List operators should be dropped.  I will either modify this or drop it, depending on how that discussion goes.  Thanks for the suggestions, I'll keep it mind int he future either way.

See: https://github.com/apache/airflow/pull/16571#discussion_r660444860

ferruzzi replied:
Looks like they are staying; corrected in coming revision to return entire list.

ferruzzi replied:
I think this should be corrected in https://github.com/apache/airflow/pull/16571/commits/fa75d5839682136fc80a2578746ecf6ef2f31822

mik-laj commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    config_text = yaml.dump(cluster_config, default_flow_style=False)
+    open(kube_config_file_location, "w").write(config_text)
```
Please use context manager. This way we are sure that the file will be closed when not in use.

ferruzzi replied:
ACK.  Good catch, thanks.

ferruzzi replied:
Fixed in 616d249

mik-laj commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,132 @@
+        eks_cluster_name,
+    ]
+    if role_arn:
+        cli_args.extend(["--role-arn", role_arn])
```
Why we need to use external tool?

mik-laj replied:
Is it impossible to generate this token using STS?
```python
import base64
import boto3
import re
from botocore.signers import RequestSigner

STS_TOKEN_EXPIRES_IN = 60

def get_bearer_token(session, cluster_id, region):
    client = session.client('sts', region_name=region)
    service_id = client.meta.service_model.service_id

    signer = RequestSigner(
        service_id,
        region,
        'sts',
        'v4',
        session.get_credentials(),
        session.events
    )

    params = {
        'method': 'GET',
        'url': 'https://sts.{}.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'.format(region),
        'body': {},
        'headers': {
            'x-k8s-aws-id': cluster_id
        },
        'context': {}
    }

    signed_url = signer.generate_presigned_url(
        params,
        region_name=region,
        expires_in=STS_TOKEN_EXPIRES_IN,
        operation_name=''
    )

    base64_url = base64.urlsafe_b64encode(signed_url.encode('utf-8')).decode('utf-8')

    # remove any base64 encoding padding:
    return 'k8s-aws-v1.' + re.sub(r'=*', '', base64_url)

# If making a HTTP request you would create the authorization headers as follows:

session = boto3.session.Session()
headers = {'Authorization': 'Bearer ' + get_bearer_token(session, 'my_cluster', 'us-east-1')}
```

mik-laj replied:
In the case of Google, we should be using normal Google tokens sometime.

``hook.get_credentials().token`` should be enough to generate a valid access token, but it is not clearly documented in public docs.

mik-laj replied:
Here is a real-world example for usage gcloud access token to access kubernetes cluster:
https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform#using-the-kubernetes-and-helm-providers

ferruzzi replied:
Using the AWS CLI was not our preference either, but had to settle for this due to other options not working correctly.  The EKS service does not support providing the token directly at all.  Searching online came up with a few workarounds, but the lesser evil was assuming that someone using an AWS service would be able to install an AWS tool.  There is also precedent for it in [Google's Kubernetes Operators](https://github.com/apache/airflow/blob/2625007c8aeca9ed98dea361ba13c2622482d71f/airflow/providers/google/cloud/operators/kubernetes_engine.py#L319) using the gcloud local script to accomplish the same thing.

I'll test using your provided example to see if that works and report back.  Dropping that dependency would be great.

mik-laj replied:
>  the lesser evil was assuming that someone using an AWS service would be able to install an AWS tool.

For me, this is very problematic for three reasons: 
- Passing credentials from Airflow to AWS CLI.
- AWS CLI is a native library because it cannot install the latest version via PIP, so in many cases it will be very difficult to use.
- Official Docker image doesn't have preinstalled AWS CLI.

mik-laj replied:
For more inspiration, you can also look at [aws-cli/awscli/customizations/eks/get_token.py](https://github.com/aws/aws-cli/blob/develop/awscli/customizations/eks/get_token.py), [kubergrunt eks token](https://github.com/gruntwork-io/kubergrunt), [kubernetes-sigs/aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator)

ferruzzi replied:
Is the gcloud tool that the GKE operator uses in that link preinstalled in the official Docker image?  If so, could the AWS tool be as well?

mik-laj replied:
No. Gcloud is not preinstalled. We have a guide that explains how to add `gcloud` to a Docker image. See: http://airflow.apache.org/docs/docker-stack/recipes.html#google-cloud-sdk-installation

ferruzzi replied:
Then I am missing what makes it ok for GKE users to have to install the gcloud cli tool but not for EKS users to have to install the aws cli tool.

mik-laj replied:
The KubernetesEngine operator requires an update. It was added when I didn't know a better way to authenticate. it should use google access-token for authentication: https://github.com/apache/airflow/pull/16571#discussion_r659255470

If you want, you can keep a dependency on AWS CLI (though I'd prefer not), but you definitely should pass credentials from Airflow to AWS CLI and that's the main issue. Now, this operator cannot be used on Cloud Composer, GKE, Astronomer, or many other environments that do not have set up the default AWS credentials in AWS CLI.

mik-laj replied:
The best way to integrate is when the token generation code would also use the Airflow hook because then we would have the greatest certainty that all credentials were passed from Airflow correctly. ;-)

ferruzzi replied:
I still agree that it would be best to drop the dependency on the AWS CLI, I was just trying to understand the difference between the two cases... and that answers it. :)

My main concern with most of the alternatives/workarounds that I found online was that if something changes in the way STS vends the tokens then we're stuck with a broken product and trying to catch up.  We can (safely???) assume that the AWS CLI would always have the correct current means to get the token.  The AWS CLI code is open source, it's trivial to just copy/paste the relevant portions here, but that is asking for a problem, later IMHO.

I'm going to try to implement your suggested `get_bearer_token` code and see where that takes me.  I appreciate your patience explaining the issue.

mik-laj replied:
> that is asking for a problem, later IMHO.

I agree that this could bring unexpected problems in the future, but having to install AWS CLI by users and integrating Airflow with AWS CLI itself can be very difficult in edge cases too. We do not have a perfect solution here. The authorization method rarely changes compared to the product API, so I believe it will be enough stable approach.

Especially since it is described in the AWS documentation.
> Amazon EKS uses IAM to provide authentication to your Kubernetes cluster through [the AWS IAM authenticator for Kubernetes](https://github.com/kubernetes-sigs/aws-iam-authenticator). 

https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html

or

> This can be used as an alternative to the aws-iam-authenticator.

https://docs.aws.amazon.com/cli/latest/reference/eks/get-token.html

In the aws-iam-authenticator documentation, we have the code snippet I cited above.
https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster

mik-laj replied:
If you are concerned about maintaining this code, why not add it to AWS SDK for Python? It could have been helpful for other people as well. Have you thought to create an internal ticket on this topic? We may not solve this problem now, but we may hope to improve it in the future.

ashb replied:
Is https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/eks.html#EKS.Client.generate_presigned_url of any help here?

ferruzzi replied:
Early tests with the proposed get_bearer_token (https://github.com/apache/airflow/pull/16571#discussion_r659252727) are promising.  I'm going to play with some edge cases and see how it holds up, but I like the direction that is taking.

In general, are we happy with the security practices used there?  Any concerns we need to pay attention to or address with that approach?

ferruzzi replied:
This should be addressed in https://github.com/apache/airflow/pull/16571/commits/a6ee2c369227212e12b8d5320c91fd63a6dc769d

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+  "eks.amazonaws.com" must be added to the Trusted Relationships
+  "AmazonEC2ContainerRegistryReadOnly" IAM Policy must be attached
+  "AmazonEKSClusterPolicy" IAM Policy must be attached
+  "AmazonEKSWorkerNodePolicy" IAM Policy must be attached
```
```suggestion
  ``ec2.amazon.aws.com`` must be in the Trusted Relationships
  ``eks.amazonaws.com`` must be added to the Trusted Relationships
  ``AmazonEC2ContainerRegistryReadOnly`` IAM Policy must be attached
  ``AmazonEKSClusterPolicy`` IAM Policy must be attached
  ``AmazonEKSWorkerNodePolicy`` IAM Policy must be attached
```
To avoid spelling check errors.

ferruzzi replied:
Fixed in 616d249

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+ - example_eks_create_cluster.py
+ - example_eks_create_cluster_with_nodegroup.py
+ - example_eks_create_nodegroup.py
+ - example_eks_pod_operator.py
```
```suggestion
 - ``example_eks_create_cluster.py``
 - ``example_eks_create_cluster_with_nodegroup.py``
 - ``example_eks_create_nodegroup.py``
 - ``example_eks_pod_operator.py``
```
To avoid spelling check errors

We do not try to teach specific examples in these guides, so we do not have to focus on them. On the other hand, when a user is interested, each example has a link that takes it to the source code.

ferruzzi replied:
Does the spell check know to ignore anything inside double back ticks?

mik-laj replied:
Yes. It ignores any code literal or code block.

ferruzzi replied:
Fixed in 616d249

mik-laj commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,132 @@
+    :param aws_region: The name of the AWS Region the EKS Cluster resides in.
+    :type aws_region: str
+    """
+    installed = which("aws")
```
How to pass the credentials to AWS CLI when we have Airflow installed at alternative cloud provider or on-permise? In the case of ``GKEStartPodOperator``, we use [the ``GoogleBaseHook.provide_authorized_gcloud`` method](https://github.com/apache/airflow/blob/2625007c8aeca9ed98dea361ba13c2622482d71f/airflow/providers/google/common/hooks/base_google.py#L483) to pass credential from Airflow to Cloud SDK.

ferruzzi replied:
The code as submitted is working on the Breeze test environment, I don't think it should matter what's hosting it.  The line you highlighted specifically is only there as a check to make sure the AWS CLI tool is installed as that is (currently) a prerequisite.

mik-laj replied:
This is important because you probably used the default credentials from the CLI in Breeze, as you were logged in there before.

In a production environment, these credentials will be provided by Airflow e.g. via the Secret backend, So AWS CLI must be able to share the same credentials. Some Airflow users have a lot of AWS credentials for different accounts and we should be able to use them correctly in this operator as well. The user should not be aware in any way that a CLI is being used to get the token.

ferruzzi replied:
Based on the above discussion, we are likely to drop the AWS CLI requirement.  If we do, does that handle the concerns here or are they still an issue?

See: https://github.com/apache/airflow/pull/16571#discussion_r659252727

ferruzzi replied:
AWS CLI dependency removed in https://github.com/apache/airflow/pull/16571/commits/a6ee2c369227212e12b8d5320c91fd63a6dc769d

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+        cluster_name: str,
+        cluster_role_arn: Optional[str] = None,
+        # A default path will be used if none is provided.
+        kube_config_file_path: Optional[str] = os.environ.get(KUBE_CONFIG_ENV_VAR, DEFAULT_KUBE_CONFIG_PATH),
```
What happens if one worker is running multiple pods? Will these credentials be overwritten? It looks like a race condition bug.

ferruzzi replied:
As it is written, the credential file will be overwritten each pod, yes.  I will look into alternatives.

ferruzzi replied:
Switched to using a named temp file and tested using a DAG which executes six pods over two clusters concurrently without issue.   I know that's a small test, but it didn't have any issues.

ferruzzi replied:
This should be addressed in https://github.com/apache/airflow/pull/16571/commits/1c9c54328c30794c824282860c7c5b11dfceca22

mik-laj commented on airflow/providers/amazon/aws/sensors/eks.py:
```diff
@@ -0,0 +1,132 @@
+from airflow.utils.decorators import apply_defaults
+
+CONN_ID = "eks"
+REGION = Session().region_name
```
Does this have the effect of making a request to the metaserver when loading the file? It seems to me that this value can also be overwritten by the connection configuration. https://github.com/apache/airflow/blob/2625007c8aeca9ed98dea361ba13c2622482d71f/airflow/providers/amazon/aws/hooks/base_aws.py#L83

ferruzzi replied:
Upon further digging, it looks like we can set the default here to None and let the base hook do its magic if the user does not provide one.  I will try that and see what happens.

Note to self:  Whatever we do here, also apply the same solution to the Operators.

ferruzzi replied:
Corrected in coming revision.

ferruzzi replied:
This should be addressed in https://github.com/apache/airflow/pull/16571/commits/5726815e0e07376449d3938a15f31fc4cbca162a

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Should we add any template fields to these operaators?

ferruzzi replied:
Sure?   Which ones do you think would be valuable to add in here?

mik-laj replied:
All fields created in ctor. We should opt-out approach use.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/4b29a39e6a9800bf69f3ebeb4cadc4a1559956ff

mik-laj commented on airflow/providers/amazon/aws/example_dags/example_eks_pod_operation.py:
```diff
@@ -0,0 +1,54 @@
+        # Optional IAM Role to assume for credentials when signing the token.
+        cluster_role_arn=ROLE_ARN,
+        image="amazon/aws-cli:latest",
+        cmds=["sh", "-c", "aws s3 mb s3://hello-world" + BUCKET_SUFFIX],
```
Can you try to avoid side-effects in tests? This way, the tests are safer for the CI environment.

ferruzzi replied:
ACK.  I'll change this to just echo to the terminal or something.

mik-laj replied:
```suggestion
        cmds=["sh", "-c", "aws sts get-caller-identity"],
```

mik-laj reviewed (changes requested):
I think we should think a little more about authorization to a cluster in ``EKSPodOperator`` operator.

ashb commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,346 @@
+
+        except ClientError as e:
+            self.log.error(e.response["Error"]["Message"])
+            raise e
```
Is this except useful? Given we are re-raising the error I would suggest not -- let this be handled by the caller and remove the try/except from all of these methods.

ferruzzi replied:
I like logging my exception messages.  You don't find them useful?

ferruzzi replied:
Removing in coming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/3d8cc9379ab9e7179f837f45bacda64fa6f2a44e

ashb commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,346 @@
+        :rtype: Dict
+        """
+        try:
+            eks_client = self.get_conn()
```
```suggestion
            eks_client = self.conn
```

`get_conn()` is qausi-deprecated, and the cached `conn` property is preferred.

ferruzzi replied:
ACK, I'll correct this in the next revision.

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+         maintained on each worker node).
+    :type aws_conn_id: str
+
+    If 'compute' is 'nodegroup', the following are required:
```
I don't think this will render correctly -- you'll have to express this in prose instead.

ferruzzi replied:
How do you feel about this:

`If compute is assigned the value of ``nodegroup``, the following are required:`

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+                    )
+                    sleep(CHECK_INTERVAL_SECONDS)
+                else:
+                    message = "Cluster is still inactive after the allocated time limit.  Aborting."
```
Should we tear it down in this case?

ferruzzi replied:
We could.   I don't suppose there is a use case where they would want it to stay, is there?  I can't think of one now that you point it out.

ferruzzi replied:
Teardown added in coming revision.

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+            # Scaling up the timeout based on the number of nodegroups that are being processed.
+            additional_seconds = 5 * 60
+            countdown = TIMEOUT_SECONDS + (nodegroup_count * additional_seconds)
+            while len(eks_hook.list_nodegroups(clusterName=self.clusterName).get('nodegroups')) > 0:
```
```suggestion
            while eks_hook.list_nodegroups(clusterName=self.clusterName).get('nodegroups'):
```

No need to check the length, just see if it's "truthy"

ferruzzi replied:
Good catch, thanks.

ferruzzi replied:
Change applied in coming revision.

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+        return eks_hook.delete_nodegroup(clusterName=self.clusterName, nodegroupName=self.nodegroupName)
+
+
+class EKSDescribeAllClustersOperator(BaseOperator):
```
Can you think of a case where having this as an _operator_ to use in a DAG is actually useful?

My gut says that all these Describe* ones don't need to be operators.

ferruzzi replied:
See [below](https://github.com/apache/airflow/pull/16571#discussion_r660444860)

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,737 @@
+        return json.dumps(nodegroup_data)
+
+
+class EKSListClustersOperator(BaseOperator):
```
Same here actually -- I can't think of a case where having this appear in a DAG is actually useful.

Did you have a use case/workflow in mind when adding these?

ferruzzi replied:
For this and above, I didn't have a particular usecase in mind, but they were API endpoints and I didn't feel it was my place to say nobody would find them useful.  You want them gone, I can remove them unless someone likes them in there.

mik-laj replied:
I guess we can use this to get the newest cluster that has a given tag to limit the use of hardcoded cluster names.

ferruzzi replied:
If nobody has any strong opinions on the matter, I'll leave them in and make the other requested changes to them.

ashb replied:
Operators should be higher level than just a single API call.

Please remove all of the Describe/List operators as I cannot see them being all that useful.

If we want to support a pattern such as Kamil has suggested then we should have a Sensor or two (EKSClusterSensor/EKSNodeGroupSesor etc) that has some higher level filtering built in to it. Such a sensor should return the cluster/node group ID (returning makes it accessible via XCom) and error if more than a single value is returned, as I think returning more than one would be a mistake?

(I'd vote for the sake of getting this merged that you remove the operators and add such sensors in a new PR, but I don't mind if you want to include those sensors in this PR)

ashb commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,133 @@
+        config_file.write(config_text)
+
+
+class UnmetDependency(BaseException):
```
```suggestion
class UnmetDependency(Exception):
```

You should almost never use BaseException.

ferruzzi replied:
ACK, will be in the next revision.  Thanks.

ashb commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,133 @@
+    installed = which("aws")
+    if installed is None:
+        message = (
+            "AWS CLI version 2 must be installed on the worker.  See: "
```
Your test doesn't check for v2 specifically -- the pip-installable v1 of awscli would also pass this check.

Do we _need_ v2 here?

ferruzzi replied:
Per Kamil's comments this may be moot.   I'll either correct or remove this depending on what comes of that convo.

see:  https://github.com/apache/airflow/pull/16571#discussion_r659252384

ferruzzi replied:
AWS CLI dependency removed in https://github.com/apache/airflow/pull/16571/commits/a6ee2c369227212e12b8d5320c91fd63a6dc769d

ashb commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,133 @@
+            "AWS CLI version 2 must be installed on the worker.  See: "
+            "https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html"
+        )
+        print(message)
```
```suggestion
```

ferruzzi replied:
I'm just gonna hang my head in shame on this one.   Sorry about that.  Will be corrected in the next revision.

ashb commented on airflow/providers/amazon/aws/utils/eks_kube_config.py:
```diff
@@ -0,0 +1,133 @@
+DEFAULT_POD_USERNAME = 'aws'
+
+
+def generate_config_file(
```
This appears to be used only in one place, EKSPodOperator, so this possibly doesn't need to be in a separate file

Perhaps this could be a method on the EksHook? (Given it connects to EKS and does describe cluster, I think that is the right place for it)

ferruzzi replied:
Yes, it is only used in the EKSPodOperator.  I had kept it to a separate utils file because it felt cleaner to me and the Hooks file was/is strictly one API endpoint, one method.  I'm happy to adjust if you feel the EKSHook class is a better home.

ferruzzi replied:
Moved into the hook module in a coming revision.

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/16571/commits/2606556296bfeaccabb513397f05a3885ac44431

ashb commented on setup.py:
```diff
@@ -500,7 +500,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'jsondiff',
     'mongomock',
-    'moto~=2.0',
+    'moto~=2.0.10',
```
```suggestion
    'moto~=2.0,>=2.0.10',
```

`~=2.0.10`  is the same as `>=2.0.10,<2.1` which is more restrictive than we want I think.

uranusjr replied:
A couple of other equivalents that may be more straightforward to understand.

* `moto>=2.0.10,<3`
* `moto>=2.0.10,==2.*`

ferruzzi replied:
Included in the next revision.

ashb commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+    EKSListClustersOperator,
+    EKSListNodegroupsOperator,
+)
+from tests.providers.amazon.aws.utils.test_eks_constants import (
```
```suggestion
from tests.providers.amazon.aws.utils.eks_constants import (
```

If that file doesn't contain any tests we shouldn't prefix it with `test_`

ferruzzi replied:
Here and below, can we compromise on `eks_test_{foo}.py`?  They don't contain tests, so I see what you are saying, but they are only used by the EKS tests and I think that should be inherent in the name.

ferruzzi replied:
Renamed in a coming revision.

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/16571/commits/66679787a678bcce6563f9a600a2ccadd3a5d269 before I found out you don't want everything squashed, apologies for the confusion/complication.

ashb commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+    SUBNETS_VALUE,
+    TASK_ID,
+)
+from tests.providers.amazon.aws.utils.test_eks_utils import convert_keys, random_names
```
Same here.

ferruzzi replied:
discussing [here](https://github.com/apache/airflow/pull/16571#discussion_r660994380)

ferruzzi replied:
Renamed in a coming revision.

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/16571/commits/66679787a678bcce6563f9a600a2ccadd3a5d269 before I found out you don't want everything squashed, apologies for the confusion/complication.

ashb reviewed (changes requested):
See individual comments (sorry, didn't do it as a review)

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+Purpose
+"""""""
+
+This example dag ``example_eks_create_cluster.py`` uses ``EKSCreateClusterOperator`` to create an Amazon
```
I've mentioned this before, but we shouldn't be covering the individual sample files.
https://github.com/apache/airflow/pull/16571#discussion_r659253383
If you are using the exampleinclude directive, each code snippet has a button that allows you to go to the source code.
<img width="931" alt="Screenshot 2021-06-30 at 01 41 49" src="https://user-images.githubusercontent.com/12058428/123881125-5f9c1480-d944-11eb-8fb4-1b07cdb82e83.png">
A similar approach is used by other organizations:
https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-s3-objects.html
https://cloud.google.com/storage/docs/reporting-changes#storage_delete_bucket_notification-cpp
<img width="905" alt="Screenshot 2021-06-30 at 01 53 06" src="https://user-images.githubusercontent.com/12058428/123881912-f3221500-d945-11eb-9de1-717deaf9a7d5.png">

ferruzzi replied:
I'm sorry, I don't quite follow what you are looking for here.  The existing doc files are all over the place as far as how detailed it should be, and I was just trying to be thorough.  Which of the doc files is our golden standard?  If you can show me an existing Airflow doc file that you like, I'll emulate that one.

mik-laj replied:
I once wrote a document on how to write documentation for operators, but now it has been deleted and I don't think I have any copy.

You can be inspired by the guides below.
http://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/natural_language.html 
http://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/datacatalog.html
http://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/mlengine.html

If you want to add something to the guides, feel free to describe it, but I'm not sure if filenames just needs to be added.

ferruzzi replied:
Should be addressed with https://github.com/apache/airflow/pull/16571/commits/5743f3469c8257948b9b73e2a7b2bdafb2a1d832

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+Airflow to Amazon Elastic Kubernetes Service (EKS) integration provides Operators to create and
+interact with the EKS clusters and compute infrastructure.
+
+ - :class:`~airflow.providers.amazon.aws.operators.eks`
```
Is it work? I think, it should be as follow:
```suggestion
 - :mod:`airflow.providers.amazon.aws.operators.eks`
```

ferruzzi replied:
Should be addressed with https://github.com/apache/airflow/pull/16571/commits/5743f3469c8257948b9b73e2a7b2bdafb2a1d832

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+  1. An Amazon EKS Cluster with underlying compute infrastructure.
+  2. The AWS CLI version 2 must be installed on the worker.
+
+  see: https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html
```
We should use descriptive links. See: https://developers.google.com/style/link-text

ferruzzi replied:
Link dropped, but will keep this in mind in the future.

mik-laj commented on docs/apache-airflow-providers-amazon/operators/eks.rst:
```diff
@@ -0,0 +1,265 @@
+Defining tasks
+""""""""""""""
+
+In the following code we delete an Amazon EKS nodegroup.
```
Ideally, you should use the class name in the operator description. The class name should be usable as a link, so you must use a reference:
``:class:`~airfflow.providers.amazon.aws.operators.EKSDeleteNodegroupOperator`. ``

ferruzzi replied:
Should be addressed with https://github.com/apache/airflow/pull/16571/commits/5743f3469c8257948b9b73e2a7b2bdafb2a1d832

ferruzzi committed:
move kubeconfig generator into `hooks/eks.py`

ferruzzi committed:
Use a tempfile to store kubeconfig data

Should Address:
https://github.com/apache/airflow/pull/16571#discussion_r659253993

ferruzzi committed:
Refactor the token generation and remove the AWS CLI dependency.

ferruzzi committed:
Move kube config into Hook class as a contextmanager

ferruzzi committed:
Rework the eks.rst doc file

ferruzzi committed:
Removed random traits in tests

ferruzzi committed:
Remove explicit region defaulting

cr https://code.amazon.com/reviews/CR-52973030

ferruzzi committed:
EKS List hooks return all results

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    :type aws_region: str
+    """
+    # Set up the client
+    session = boto3.Session(region_name=aws_region, profile_name=aws_profile)
```
We should use the Hook to create a session because this will provide a uniform way to manage credentials and will allow us to use this operator in any environment i.e. on-premiss, on GCP, on Cloud Composer and others.  Now, this operator can only be used in AWS environment, because it supports only [Default Credential Provider Chain](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html).

ferruzzi replied:
ACK.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    pod_username: Optional[str] = DEFAULT_POD_USERNAME,
+    pod_context: Optional[str] = DEFAULT_CONTEXT_NAME,
+    aws_region: Optional[str] = None,
+) -> str:
```
```suggestion
def generate_config_file(
    eks_cluster_name: str,
    eks_namespace_name: str,
    pod_name: str,
    pod_username: Optional[str] = DEFAULT_POD_USERNAME,
    pod_context: Optional[str] = DEFAULT_CONTEXT_NAME,
    aws_conn_id: str = 'aws_default'
) -> str:
```

mik-laj replied:
We should accept conn_id to ensure unify credential managementt.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    """
+    # Set up the client
+    session = boto3.Session(region_name=aws_region, profile_name=aws_profile)
+    eks_client = session.client("eks")
```
```suggestion
    hook = AwsBaaseHook(aws_conn_id=aws_conn_id, clien_type='eks')
    session, _ = hook._get_credentials()
    eks_client = hook.conn
```

mik-laj replied:
We should accept conn_id to ensure unify credential management.

ferruzzi replied:
This may be going back towards your previous request/point, but would a more simple alternative to this be to move the `generate_config_file` and `_get_bearer_token` methods into the `EKSHook` class itself?  i think that was your initial suggestion and I missed the mark on this.  I think if we do it that way, then the Operator creates the hook as usual and calls it with `eks_hook.generate_config_file()`, then those three lines you highlighted are just `eks_client = self.conn` like any of the hook methods, and inside `_get_bearer_token`, we can use:

```
        session = self.get_session()
        client = self.conn
```

and I _believe_ that solution would share the conn_id between them freely instead of having to pass it around?  Is that right?

mik-laj replied:
A good idea. Writing this function as a hook method will follow how GoogleBaseHook works. In  `GoogleBaseHook`, we have `provide_gcp_credential_file_as_context` method, so we can create a similar method here e.g. `provide_eks_credential_file_as_context`.
https://github.com/apache/airflow/blob/83cb237031dfe5b7cb5238cc1409ce71fd9507b7/airflow/providers/google/common/hooks/base_google.py#L448

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+
+    # Set the filename to something which can be found later if needed.
+    filename_prefix = KUBE_CONFIG_FILE_PREFIX + pod_name
+    with tempfile.NamedTemporaryFile(prefix=filename_prefix, mode='w', delete=False) as config_file:
```
We should delete the credentials when they are not used. To achieve this, we should make the following changes:
- convert this method into context manager i.e. add [`contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) decorator, yield on a new file name (before we should flush content also), update usages,
- remove delete parameter in this method invocation
- remove pod_name parameters from this method.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    with tempfile.NamedTemporaryFile(prefix=filename_prefix, mode='w', delete=False) as config_file:
+        config_file.write(config_text)
+
+    return config_file.name
```
```suggestion
    with tempfile.NamedTemporaryFile(prefix=KUBE_CONFIG_FILE_PREFIX, mode='w') as config_file:
        config_file.write(config_text)
        config_file.flush()
        yield config_file
```

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    return config_file.name
+
+
+def _get_bearer_token(session: boto3.Session, cluster_id: str, aws_region: str) -> str:
```
Can we replace `aws_region` parameter with [boto3.Session.region_name](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.region_name)? It does not seem necessary to me to pass the same value in two parameters.

mik-laj replied:
Can you give the method a more descriptive name or add method documentation that explains what the purpose of this token is? I am afraid it is not legible yet, but it may not be very clear in the future. It is also worth adding a short note about this algorithm of this token, e.g. source (aws-iam-authenticator)

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,709 @@
+        self.region = region
+
+    def execute(self, context):
+        self.config_file = generate_config_file(
```
We should use the context manager here to ensure that the file is deleted when not in use.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,709 @@
+        pod_name: Optional[str] = DEFAULT_POD_NAME,
+        pod_username: Optional[str] = DEFAULT_POD_USERNAME,
+        aws_profile: Optional[str] = None,
+        region: Optional[str] = None,
```
We should read these parameters from connection to provide unified user experience. No other operator supports such a session configuration parameter as an operator parameter.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    :type aws_profile: str
+    :param pod_username: The username under which to execute the pod.
+    :type pod_username: str
+    :param pod_context: The name of the context access parameters to use.
```
The value of this parameter has no effect on behavior. This is just a temporary profile name. We can use a different name each time or use a constant string and the user won't notice the difference because we only have one context at a time.

ferruzzi replied:
Alright, sounds good.  I had thought there may be value in the user being able to dictate the name of the pod and be able to look up the file in the future, but if it is being deleted in the context manager, then you are correct and this is not useful.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    :type eks_cluster_name: str
+    :param eks_namespace_name: The namespace to run within kubernetes.
+    :type eks_namespace_name: str
+    :param pod_name: The unique name to give the pod.  Used as an identifier in the config filename.
```
Can we generate this pod name for each run?

ferruzzi replied:
Absolutely, that would be easy enough.   Is there no use case where the user would want to know the pod name?

mik-laj replied:
Accessing the pod name in the operator makes sense, but it doesn't make sense to strictly associate the configuration file with a specific pod name. One configuration file can be used to create multiple pods or perform other operations, such as creating a secret.

mik-laj replied:
You can keep this argument if you want, but it should have a generic name like `tmpfile_prefix` and be optional. Most users don't pay much attention to names for temporary files, so we shouldn't force it to pass a value for this argument.

ferruzzi replied:
If the file is being deleted as you requested, then this isn't really relevant anymore.  I'll change this to just use the default NamedTemporaryFile naming convention instead of overwriting that.

ferruzzi replied:
Should be addressed by https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+    :param eks_cluster_name: The name of the cluster to create the EKS Managed Nodegroup in.
+    :type eks_cluster_name: str
+    :param eks_namespace_name: The namespace to run within kubernetes.
+    :type eks_namespace_name: str
```
I don't understand the name of this parameter. Did you mean kubernetes(`k8s`) namespace or does the EKS service have different namespaces?

ferruzzi replied:
renamed in https://github.com/apache/airflow/pull/16571/commits/8687b736994d1911011dd8bd2b2903d82ad30a8c

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,461 @@
+            {
+                "name": pod_username,
+                "user": {
+                    "token": token,
```
This token is only valid for an hour, but this operator can be used to run a longer job. We should take care of its rotation of this key to ensure it's valid.

Unfortunately, it won't be trivial as we have to write a new [out-of-tree client
    authentication provider](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/kubectl-exec-plugins.md). To do this, you should write a new CLI tool that will generate a new credential.

Some tips on adding a new CLI tool to providers:
- You cannot modify the Airflow core so that this operator can be used on different Airflow versions and not create a dependency between the vendor and the core. You cannot modify any file in `airflow.cli` package.
- You can create a new [executable module](https://docs.python.org/3/tutorial/modules.html#executing-modules-as-scripts). To run this module, it is best to use python module resolution i.e. you should use [python -m](https://docs.python.org/3/using/cmdline.html#cmdoption-m). BTW, To run airflow, you can use `airflow` or `python -m airflow`. See: https://github.com/apache/airflow/pull/7808
- The `python`/`python3` may not refer to the current interpreter. You should [`sys.executable`](https://docs.python.org/3/library/sys.html#sys.executable) instead.

For example, see: [`airflow.providers.google.common.utils.id_token_credentials`](https://github.com/apache/airflow/blob/866a601b76e219b3c043e1dbbc8fb22300866351/airflow/providers/google/common/utils/id_token_credentials.py#L24])

mik-laj replied:
For more info about kubectl exec auth, see: 
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins

mik-laj commented on airflow/providers/amazon/aws/example_dags/example_eks_pod_operation.py:
```diff
@@ -0,0 +1,54 @@
+####
+
+CLUSTER_NAME = 'existing-cluster-with-nodegroup-ready-for-pod'
+BUCKET_SUFFIX = datetime.now().strftime("-%Y%b%d-%H%M").lower()
```
```suggestion
```

ferruzzi replied:
This is dropped already in a coming revision; a previous comment asked that sample dags not generate external artifacts; this dag now just performs `cmds=["sh", "-c", "ls"]`.

mik-laj commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,452 @@
+        :param pod_username: The username under which to execute the pod.
+        :type pod_username: str
+        :param pod_context: The name of the context access parameters to use.
+        :type pod_context: str
```
Can you docs about return values? It is not obvious that this is a name for a temporary file.

ferruzzi replied:
Nothing here has to do with a filename; it is a randomly generated name.  Those values are passed into the kubeconfig file.

mik-laj commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+
+class TestEKSListNodegroupsOperator(unittest.TestCase):
+    def setUp(self) -> None:
+        self.cluster_name: str = random_names()
```
Is it needed? Why can't you use a constant value?

ferruzzi replied:
Removed in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+        self, mock_describe_nodegroup, mock_list_nodegroups
+    ):
+        nodegroup_names: List[str] = NAME_LIST
+        cluster_name: str = random_names()
```
Is it needed? Can you use a constant value?

ferruzzi replied:
Removed in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/hooks/test_eks.py:
```diff
@@ -0,0 +1,765 @@
+    def test_delete_nodegroup_throws_exception_when_nodegroup_not_found(
+        self, nodegroup_builder, initial_batch_size: int = BatchCountSize.SMALL
+    ) -> None:
+        eks_hook, generated_test_data = nodegroup_builder(count=initial_batch_size)
```
Is it needed? Can you use a fixed value here?

mik-laj commented on tests/providers/amazon/aws/utils/eks_test_constants.py:
```diff
@@ -0,0 +1,256 @@
+KUBERNETES_NETWORK_CONFIG_KEY: str = "kubernetesNetworkConfig"
+KUBERNETES_NETWORK_CONFIG_VALUE: Dict = {"serviceIpv4Cidr": "172.20.0.0/16"}
+
+LABELS_KEY: str = "labels"
```
This variable has only one usage. Can we inline it to simplify this code

ferruzzi replied:
Collapsed key/value and tuple declarations in coming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/utils/eks_test_constants.py:
```diff
@@ -0,0 +1,256 @@
+DEFAULT_MAX_RESULTS = 100
+FROZEN_TIME = "2013-11-27T01:42:00Z"
+PACKAGE_NOT_PRESENT_MSG = "mock_eks package not present"
+PARTITIONS: List[str] = Session().get_available_partitions()
```
Is it needed?  Can you use a fixed value here?

ferruzzi replied:
Removed in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/utils/eks_test_constants.py:
```diff
@@ -0,0 +1,256 @@
+FROZEN_TIME = "2013-11-27T01:42:00Z"
+PACKAGE_NOT_PRESENT_MSG = "mock_eks package not present"
+PARTITIONS: List[str] = Session().get_available_partitions()
+REGION: str = Session().region_name
```
Is it needed?  Can you use a fixed value here?

ferruzzi replied:
Removed both in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+class TestEKSDeleteNodegroupOperator(unittest.TestCase):
+    def setUp(self) -> None:
+        self.cluster_name: str = random_names()
+        self.nodegroup_name: str = random_names()
```
Is it needed?  Can you use a fixed value here?

ferruzzi replied:
Removed in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

mik-laj commented on tests/providers/amazon/aws/operators/test_eks.py:
```diff
@@ -0,0 +1,299 @@
+
+class TestEKSDeleteNodegroupOperator(unittest.TestCase):
+    def setUp(self) -> None:
+        self.cluster_name: str = random_names()
```
Is it needed to generate random value?  Can you use a fixed value here?

ferruzzi replied:
Removed in upcoming revision.

ferruzzi replied:
Corrected in https://github.com/apache/airflow/pull/16571/commits/76c7ad635948191764a323a9be03fa965fc91ff5

ferruzzi committed:
Implemented Jinja templates for operators

- Added jinja template fields
- Refactored fields to snake_case since they are now exposed
- Removed a couple of straggling pylint instructions; pylint is no longer used

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,709 @@
+
+CHECK_INTERVAL_SECONDS = 15
+TIMEOUT_SECONDS = 25 * 60
+CONN_ID = "eks"
```
All AWS should use ``aws_default`` as the default connection ID. See: http://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html

ferruzzi replied:
addressed in https://github.com/apache/airflow/pull/16571/commits/08c2a3480609d8c0cd902837d1ccb420cc93b4cf

mik-laj commented on airflow/providers/amazon/aws/sensors/eks.py:
```diff
@@ -0,0 +1,129 @@
+        *,
+        cluster_name: str,
+        target_state: Optional[str] = TARGET_STATE,
+        conn_id: Optional[str] = CONN_ID,
```
```suggestion
        aws_conn_id: Optional[str] = CONN_ID,
```
All AWS operators use this name. For consistency, we should follow this convention.

ferruzzi replied:
gah.  I caught/corrected that in the Operators but missed it in the Sensors, thanks.

ferruzzi committed:
Remove try/log blocks from hooks

ferruzzi committed:
conn_id refactor

- Refactored using IDE magic:
 - any field named `self.conn_id` is now `self.aws_conn_id`
 - any param named `conn_id` is now `aws_conn_id`
 - any constant named `CONN_ID` is now `DEFAULT_CONN_ID`
- Sensors were missing template fields, added those.

ferruzzi committed:
Implemented EKS system tests

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,797 @@
+            else:
+                message = "Creating an EKS Managed Nodegroup requires nodegroup_role_arn to be passed in."
+                self.log.error(message)
+                raise AttributeError(message)
```
We don't need the log and to raise an error -- by raising the error the message will bubble up anyway, so logging here would just lead to the message appearing twice.

I have also "inverted" the check to have the main flow be "un-intendented" (A _rough_ rule I follow is to try and avoid having code in a block, with the thinking being that the deeper in a code the harder it is to follow.)

```suggestion
            if not nodegroup_role_arn:
                raise ValueError("Creating an EKS Managed Nodegroup requires nodegroup_role_arn to be passed in.")
            self.nodegroup_role_arn = nodegroup_role_arn
```

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/16571/commits/4d59785e903699d74d884a9137197baef3fe0fde

ashb commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,797 @@
+        nodegroup_name: Optional[str] = None,
+        nodegroup_role_arn: Optional[str] = None,
+        compute: Optional[str] = DEFAULT_COMPUTE_TYPE,
+        aws_conn_id: Optional[str] = DEFAULT_CONN_ID,
```
This shouldn't be optional -- passing `None` isn't valid is it?

```suggestion
        aws_conn_id: str = DEFAULT_CONN_ID,
```

ashb replied:
(This applies to all operators added here)

mik-laj replied:
Connection I'D is optional for all AWS operators. See: https://github.com/apache/airflow/pull/8534

ashb replied:
Ah cool. I just assumed it used the connection named "aws_default", which was "empty".

ferruzzi committed:
Fixed `nextToken` final result bug

ferruzzi committed:
Remove List and Describe Operators and supporting code.

ferruzzi commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,797 @@
+                    self.log.error(message)
+                    # If there is something preventing the cluster for activating, tear it down and abort.
+                    eks_hook.delete_cluster(name=self.cluster_name)
+                    raise RuntimeError(message)
```
@ashb  - Working on removing the redundant catch/log/except blocks. How do you feel about this one? Should I drop this log as well, or is that reasonable?  (here and a very similar use below, on/around L300 )

ashb replied:
How long does deleting the cluster take? if it's more than a few seconds it's probably useful to keep both here, but on L300 the log and the exception are right next to each other, so we should remove the log form there.

ferruzzi replied:
It's in the 15-20 minute range, in my experience.

ashb replied:
Oh one more thing: is `delete_cluster` sync or async? If it's async (and we just wait for the request) then we don't need the log.

ferruzzi replied:
I thought I replied here but apparently not.  IKt's async, it sends the request to the API then moves on.  I've removed this logging in https://github.com/apache/airflow/pull/16571/commits/4d59785e903699d74d884a9137197baef3fe0fde

ferruzzi committed:
Remove some nesting and some unnecessary logging before raising an exception

ferruzzi committed:
Use the force

ferruzzi commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -0,0 +1,797 @@
+            )
+            for group in nodegroups:
+                eks_hook.delete_nodegroup(clusterName=self.cluster_name, nodegroupName=group)
+
```
@mik-laj @ashb @o-nikolas  - I added this as a convenience but now I'm having second thoughts and want your opinion.  By default, the API will not allow you to delete a Cluster which has attached Nodegroups.  I added this block to check for nodegroups and delete them for you rather than throwing an exception, but I now wonder if I should put that behind a `force delete` flag rather than circumvent the system behavior by default?

ashb replied:
Yeah, you are probably right that having it behind a `force=True` flag would be better.

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/16571/commits/553e4835ea846c447f21398effefd0d1dbc13c4c

ferruzzi committed:
Additional jinja templating

ferruzzi committed:
Improved docs and samples

zkan commented on airflow/providers/amazon/aws/hooks/eks.py:
```diff
@@ -0,0 +1,420 @@
+        )
+        return response
+
+    def describe_cluster(self, name: str, verbose: Optional[bool] = False) -> Dict:
```
I'm very new to type hints and wonder if it's more common to use `Optional` than this:
```suggestion
    def describe_cluster(self, name: str, verbose: bool = False) -> Dict:
```

May I have your advice on this, please? Thanks in advance! :)

mik-laj replied:
> You can use the Optional type modifier to define a type variant that allows None, such as Optional[int] (Optional[X] is the preferred shorthand for Union[X, None]):

https://mypy.readthedocs.io/en/latest/kinds_of_types.html#optional-types-and-the-none-type

I agree. We should use `bool` type here.

ferruzzi replied:
Sorry for the delay, I'll make the change.

ferruzzi replied:
Found a few other places where I think I misused Optional, I'll be dropping a commit in here with those corrections later today or tomorrow.

ferruzzi replied:
@zkan - should be addressed in https://github.com/apache/airflow/pull/16571/commits/d63063932c4e42432a039fc69d9c592ce08b95c6

ferruzzi committed:
Corrected some misused Optionals.

ferruzzi committed:
Doc formatting fix

ashb reviewed (commented):
Code/features looks good now, the integration logo appears to be empty though?

ferruzzi committed:
Corrected logo

ferruzzi committed:
Corrected logo - background and size
